docs: create a separate effect guide (#64323)
angular/angular: adev/src/content/guide/signals/effect.md

## Effects

Signals are useful because they notify interested consumers when they change. An **effect** is an operation that runs whenever one or more signal values change. You can create an effect with the `effect` function:

```ts
effect(() => {
  console.log(`The current count is: ${count()}`);
});
```

Effects always run **at least once.** When an effect runs, it tracks any signal value reads. Whenever any of these signal values change, the effect runs again. Similar to computed signals, effects keep track of their dependencies dynamically, and only track signals which were read in the most recent execution.

> Effects always execute **asynchronously**, during the change detection process.

Author: effects scheduling is alaways tied to CD/tick. root effect before any component is checked, view effects before each corresponding component. I'm not sure how much is worth mentionning

## Effects

Signals are useful because they notify interested consumers when they change. An **effect** is an operation that runs whenever one or more signal values change. You can create an effect with the `effect` function:

```ts
effect(() => {
  console.log(`The current count is: ${count()}`);
});
```

Effects always run **at least once.** When an effect runs, it tracks any signal value reads. Whenever any of these signal values change, the effect runs again. Similar to computed signals, effects keep track of their dependencies dynamically, and only track signals which were read in the most recent execution.

Effects always execute **asynchronously**, during the change detection process.

### Use cases for effects

> Effects are rarely needed in most application code, but may be useful in specific circumstances. Here are some examples of situations where an `effect` might be a good solution:

Reviewer: We should steal some text from https://github.com/angular/angular/blob/9ffc83e163f681b92a409eb5cac0837dbc865f08/adev/src/content/tutorials/signals/steps/10-reacting-to-signal-changes-with-effect/README.md for discouraging overuse of `effect`. In particular, the guidance about it being the last thing you reach for and the examples of specific situations when to avoid `effect`. 

(as well as examples for when `effect` is appropriate) 
(not necessarily for the docs, but I kind of like the phrase "There are no situations where `effect` is _good_, only situations where it is _appropriate_.")

```ts
@Component({...})
export class EffectiveCounterComponent {
  readonly count = signal(0);
  private injector = inject(Injector);

  initializeLogging(): void {
    effect(() => {
      console.log(`The count is: ${this.count()}`);
    }, {injector: this.injector});
  }
}
```

### When to Use (and Not to Use) Effects

Effects are a reactivity primitive, but they are not always the right solution. It's important to understand when to use them and when to reach for other tools like `computed`, `linkedSignal` or `resource`.

#### Good Use Cases for Effects

<!-- TODO -->

#### When to Avoid Effects:

<!-- TODO -->

### Execution of effects

<!-- Root vs View effects -->

### Destroying effects

When you create an effect, it is automatically destroyed when its enclosing context is destroyed. This means that effects created within components are destroyed when the component is destroyed. The same goes for effects within directives, services, etc.

An `effect` can be created in 2 different context that will affect when it's destroyed:

- A "View effect" is created in the context of a component, will be destroyed when the component is destroyed.

> - A "Root effect" is created in a root provided service, and if destroyed when the application itself is destroyed.

Reviewer: ```suggestion
- A "View effect" is created in the context of a component, is destroyed when the component is destroyed.
- A "Root effect" is created in a root provided service, and is destroyed when the application is destroyed.
```

Effects return an `EffectRef` that you can use to destroy them manually, by calling the `.destroy()` method. You can combine this with the `manualCleanup` option to create an effect that lasts until it is manually destroyed. Be careful to actually clean up such effects when they're no longer required.

### Effect cleanup functions

Effects might start long-running operations, which you should cancel if the effect is destroyed or runs again before the first operation finished. When you create an effect, your function can optionally accept an `onCleanup` function as its first parameter. This `onCleanup` function lets you register a callback that is invoked before the next run of the effect begins, or when the effect is destroyed.

```ts
effect((onCleanup) => {
  const user = currentUser();

  const timer = setTimeout(() => {
    console.log(`1 second ago, the user became ${user}`);
  }, 1000);

  onCleanup(() => {
    clearTimeout(timer);
  });
});
```

## Side effects on DOM elements

> The `effect` function is a general-purpose tool for running code in reaction to signal changes. However, it runs _before_ the components are checked and the DOM has been updated. This is great for many use cases, but sometimes you need to run code _after_ the DOM has been updated. For example, you might need to manually inspect or modify the DOM, or integrate a 3rd-party library that requires direct DOM access.

Reviewer: ```suggestion
The `effect` function is a general-purpose tool for running code in reaction to signal changes. However, it runs _before_ the Angular updates the DOM. In some situations, you may need to manually inspect or modify the DOM, or integrate a 3rd-party library that requires direct DOM access.
```

- `write`: Use this phase to write to the DOM. **Never** read from the DOM in this phase.
- `mixedReadWrite`: Use this phase to read from and write to the DOM simultaneously. Never use this phase if it is possible to divide the work among the other phases instead.
- `read`: Use this phase to read from the DOM. **Never** write to the DOM in this phase.

Using these phases helps prevent layout thrashing and ensures that your DOM operations are performed in a safe and efficient manner.

You can specify the phase by passing an object with a `phase` property to `afterRender` or `afterNextRender`:

```ts
afterRenderEffect({
    earlyRead: (cleanupFn) => { /* ... */ },
    write: (previousPhaseValue, cleanupFn) => { /* ... */ },
    mixedReadWrite: (previousPhaseValue, cleanupFn) =>  { /* ... */ },
    read: (previousPhaseValue, cleanupFn) =>  { /* ... */ },
});
```

CRITICAL: If you don't specify the phase for the callback, `afterRenderEffect` will run it during the `mixedReadWrite` phase. By doing so you risk significant performance degradation.

#### Phase executions

> The first phase callback to run as part of this spec will receive no parameters. Each subsequent phase callback in this spec will receive the return value of the previously run phase callback as a Signal. This can be used to coordinate work across multiple phases.

Reviewer: ```suggestion
The `earlyRead` phase callback receives no parameters. Each subsequent phase receives the return value of the previous phase's callback as a Signal. You can use this to coordinate work across phases.
```

> Effects run in the following phase order, only when dirty through signal dependencies:

Reviewer: ```suggestion
Effects run in the following phase order:
```
IMO the guide already covered how effects run based on signal deps changing
